Sources/Action/InputSubject.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+import RxSwift
+
+/// A special subject for Action.inputs. It only emits `.next` event.
+public class InputSubject<Element>: ObservableType, Cancelable, SubjectType, ObserverType {
+
+    public typealias E = Element
+    typealias Key = UInt
+
+    /// Indicates whether the subject has any observers
+    public var hasObservers: Bool {
+        _lock.lock()
+        let count = _observers.count > 0
+        _lock.unlock()
+        return count
+    }
+    
+    // state
+    private let _lock = NSRecursiveLock()
+    private var _nextKey: Key = 0
+    private var _observers: [Key: (Event<Element>) -> ()] = [:]
+    private var _isDisposed = false
+
+    /// Indicates whether the subject has been isDisposed.
+    public var isDisposed: Bool {
+        return _isDisposed
+    }
+
+    /// Creates a subject.
+    public init() {
+        #if TRACE_RESOURCES
+        _ = Resources.incrementTotal()
+        #endif
+    }
+
+    /// Notifies all subscribed observers abount only `.next` event.
+    ///
+    /// - parameter event: Event to send to the observers.
+    public func on(_ event: Event<Element>) {
+        _lock.lock()
+        switch event {
+        case .next(_) where !_isDisposed:
+            _observers.values.forEach { $0(event) }
+        default:
+            break
+        }
+        _lock.unlock()
+    }
+
+    /**
+     Subscribes an observer to the subject.
+
+     - parameter observer: Observer to subscribe to the subject.
+     - returns: Disposable object that can be used to unsubscribe the observer from the subject.
+     */
+    public func subscribe<O: ObserverType>(_ observer: O) -> Disposable where O.E == Element {
+        _lock.lock()
+
+        if _isDisposed {
+            observer.on(.error(RxError.disposed(object: self)))
+            return Disposables.create()
+        }
+        
+        let key = _nextKey
+        _nextKey += 1
+        _observers[key] = observer.on
+        _lock.unlock()
+
+        return Disposables.create { [weak self] in
```

Comment:
@sunshinejr `_isDisposed` is set to `false` on line 22. But my guess is that you have meant setting it to `true`. The thing is that here disposable for *subscription* is created, not for the `InputSubject` itself.